### DISCUSSION TITLE
How can I get console/diagnostic output from the new v3 `AssemblyFixture`?

### DISCUSSION THREAD

**Discussion by timabell** [xunit/xunit#3015] [Question] [answered]
I'm trying out the new [AssemblyFixture](https://xunit.net/docs/shared-context#assembly-fixture) in the preview release of xunit v3.

I tried giving the assembly fixture this constructor:
```csharp
public AssemblyFixtureExample(IMessageSink output)
```
as per https://xunit.net/docs/capturing-output
but that results in
```
Assembly fixture type 'AssemblyFixtureExample' had one or more unresolved constructor arguments: IMessageSink output
```

and I tried giving the assembly fixture this constructor:
```csharp
public AssemblyFixtureExample(IMessageSink output)
```
as per https://xunit.net/docs/capturing-output
but that results in
```
Assembly fixture type 'AssemblyFixtureExample' had one or more unresolved constructor arguments: ITestOutputHelper output
```

Is there any way of getting some kind of console output from the assembly fixture methods?

Context: Working on https://github.com/timabell/CollectionFixtureXUnitExample/tree/v3again/V3AssemblyFixture

**Comment by bradwilson** [comment, marked as answer]
We have not updated the documentation yet for v3.

The equivalent to send a diagnostic message in v3 is to just call `TestContext.Current.SendDiagnosticMessage`. No more need to inject the message sink.

**Reply by timabell** [reply]
Hi thanks for the tip, I've finally found some time to try it out and I can't seem to get any output from that either on the console or in Rider's test runner.

Current attempt in this branch https://github.com/timabell/CollectionFixtureXUnitExample/tree/SendDiagnosticMessage/V3AssemblyFixture

I only see the output helper output, and not the diagnostic output:

```
		_output.WriteLine($"- Running Assembly {nameof(TestClass1)}.{nameof(Test1)}"); # works in test method
		TestContext.Current.SendDiagnosticMessage("diag Running Assembly {nameof(TestClass1)}.{nameof(Test1)}"); # can't see this
```

![image](https://github.com/user-attachments/assets/41482053-dc0b-4ac3-a583-87486f337a76)

![image](https://github.com/user-attachments/assets/bc57204e-fa73-4a36-b5ad-48c2db70dfc2)

**Reply by bradwilson** [reply]
Diagnostic messages are not shown by default: https://xunit.net/docs/configuration-files#diagnosticMessages

**Reply by timabell** [reply]
Sorry, I should have mentioned, I had already found that setting and had created the mentioned file - https://github.com/timabell/CollectionFixtureXUnitExample/blob/SendDiagnosticMessage/V3AssemblyFixture/xunit.runner.json#L3

**Reply by bradwilson** [reply]
In Visual Studio, they're visible in the Tests panel of the Output window:

![image](https://github.com/user-attachments/assets/1e15a729-6dec-41cd-b6d7-90902d970ece)

If they're not visible in Rider, then that's an issue for the Rider team.

**Reply by Jarryd-Lansdell** [reply]
@timabell I checked out your given code and tried to run it and I didn't see the log when using the cli(your debug-run.sh), did you get this to work?
I've tried for a while and never got this to work using assembly fixture. Even the example given by @bradwilson the code is in the test, not in the assembly fixture.

**Reply by bradwilson** [reply]
<img width="1652" height="1314" alt="image" src="https://github.com/user-attachments/assets/130c9904-2356-4faa-84d1-5b941a77f9fe" />

**Reply by bradwilson** [reply]
<img width="1630" height="1378" alt="image" src="https://github.com/user-attachments/assets/02073f2c-8877-48e8-93b5-9f79f5bcc811" />

**Comment by timabell** [comment]
Should the diagnostic messages show when run from the dotnet cli?

Do you have any links you could share on how that diagnostic messages thing works so I can look into it? I'm assuming that xunit3 is hooking in to some well known API?

**Reply by bradwilson** [reply]
Add `--logger "console;verbosity=normal"` to "un-hide" xUnit.net's messages (VSTest, aka `dotnet test`, hides them by default).
